File: rides-service/src/main/java/com/example/rides_service/config/RetrieveMessageErrorDecoder.java
package com.example.rides_service.config;

import com.example.rides_service.util.ExceptionMessages;
import feign.Response;
import feign.codec.ErrorDecoder;
import jakarta.persistence.EntityNotFoundException;
import lombok.extern.slf4j.Slf4j;
import org.apache.coyote.BadRequestException;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.stream.Collectors;

@Slf4j
public class RetrieveMessageErrorDecoder implements ErrorDecoder {

    private ErrorDecoder errorDecoder = new Default();

    @Override
    public Exception decode(String s, Response response) {
        String errorMessage = ExceptionMessages.UNKNOWN_ERROR.toString();
        if (response.body() != null) {
            try (InputStream bodyIs = response.body().asInputStream()) {
                errorMessage = new BufferedReader(new InputStreamReader(bodyIs))
                        .lines()
                        .collect(Collectors.joining("\n"));
            } catch (IOException e) {
                log.error("Error reading response body: {}", e.getMessage());
                return new Exception(ExceptionMessages.UNABLE_TO_READ_ERROR_RESPONSE.format());
            }
        }
        return switch (response.status()) {
            case 400 -> new BadRequestException(errorMessage);
            case 404 -> new EntityNotFoundException(errorMessage);
            default -> errorDecoder.decode(s, response);
        };
    }
}

File: rating-service/src/main/kotlin/com/modsen/rating/client/RetrieveMessageErrorDecoder.kt
package com.modsen.rating.client

import com.modsen.rating.util.ExceptionMessages
import feign.Response
import feign.codec.ErrorDecoder
import jakarta.persistence.EntityNotFoundException
import java.io.IOException
import org.apache.coyote.BadRequestException
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component

@Component
class RetrieveMessageErrorDecoder : ErrorDecoder {

    private val log = LoggerFactory.getLogger(com.modsen.rating.client.RetrieveMessageErrorDecoder::class.java)
    private val defaultErrorDecoder = ErrorDecoder.Default()

    override fun decode(methodKey: String, response: Response): Exception {
        var errorMessage = ExceptionMessages.UNKNOWN_ERROR

        response.body()?.let { responseBody ->
            try {
                errorMessage = responseBody.asInputStream().bufferedReader().use { it.readText() }
            } catch (e: IOException) {
                log.error("Error reading response body: {}", e.message)
                return Exception(ExceptionMessages.UNABLE_TO_READ_ERROR_RESPONSE)
            }
        }

        return when (response.status()) {
            400 -> BadRequestException(errorMessage)
            404 -> EntityNotFoundException(errorMessage)
            else -> defaultErrorDecoder.decode(methodKey, response)
        }
    }

}